ADR/0027-tekton-results-naming-convention.md

```diff
@@ -0,0 +1,150 @@
+    "note": {
+      "type": "string",
+    },
+    "failures": {
+      "type": "integer",
+      "minimum": 0
+    },
+    "warnings": {
+      "type": "integer",
+      "minimum": 0
+    }
+  },
+  "required": ["result", "timestamp", "successes", "failures", "warnings"]
+}
+```
+
+### Results for scan-like tasks
+Scan-like tasks are those that search for issues within the code base or artifact whose results raise concerns (vulnerabilities) according to some internal severity or classification metric. These results are expressly different from test-like results as there may need to be additional criteria applied to the results to determine what vulnerabilities are permissable for the final artifact.
+
+Scan-like results should still use the **TEST_OUTPUT** result for indicating whether the scan has run successfully or if it has errored, but this output should not indicate a failure.
+
+To display count of found vulnerabilities and make it easy to understand and evaluate the state of scanned content, the additional output of scan-like tasks will be provided in a minimized [Tekton result](https://tekton.dev/docs/pipelines/tasks/#emitting-results) in JSON format listing. The name of the result will be **SCAN_OUTPUT**.
+
+While the vulnerability classifications should remain consistent in order to enable easier extensions into other AppStuido components (namely the user interface and enterprise contract), it is the responsibility of every scan-like task to inform the user about what criteria fits each vulnerability classification used. The vulnerability classifications from most important/severe to least are **critical**, **high**, **medium**, and **low**.
+
+#### Tekton Result Format for `SCAN_OUTPUT`
+The Test output of the Tekton result **SCAN_OUTPUT** will be a JSON object that includes the following information about the found vulnerabilities. While the vulnerability classification may vary based on the scanner used in the task, an example description of vulnerability secerity ratings can be fount at the [Red Hat Vulnerability documentation](https://access.redhat.com/articles/red_hat_vulnerability_tutorial).
+
+```
+{
+  "vulnerabilities": {
+    "critical": 1,
+    "high": 0,
+    "medium": 1,
+    "low":0
+  }
+}
+```
+
+#### `SCAN_OUTPUT` Schema Validation
+The output of the Tekton result **SCAN_OUTPUT** can be validated using the following schema:
+
+```
+{
+  "$schema": "http://json-schema.org/draft/2020-12/schema#",
+  "type": "object",
+  "properties": {
+    "vulnerabilities": {
+      "type": "object",
+      "properties": {
+        "critical": {
+          "type": "integer",
+          "minimum": 0
+        },
+        "high": {
+          "type": "integer",
+          "minimum": 0
+        },
+        "medium": {
+          "type": "integer",
+          "minimum": 0
+        },
+        "low": {
+          "type": "integer",
+          "minimum": 0
+        }
+      },
+      "required": ["critical", "high", "medium", "low"]
```

Comment:
@arewm Do you want to consider if there's a fix available as part of this proposal (https://issues.redhat.com/browse/RHTAP-1015)? or is that too specific for this moment in time?

Reply:
Clair and roxctl both have the fixable classification. We can propose something to meet that request in this ADR change or it can be left out.

If we are to try to encapsulate the fixable/unpatched concept into this return value, I would probably suggest a format like

```
{
  "vulnerabilities": {
    "critical": 1,
    "high": 0,
    "medium": 1,
    "low":0
  },
  "unpatched_vulnerabilities": {
    "critical": 0,
    "high": 1,
    "medium": 0,
    "low":1
  }
}
```

where the backwards-compatible case would default to those that are fixable and the unfixable ones have the new classification which _can_ be introduced in policies if desired.

Reply:
@MartinBasti, FYI

Reply:
+1

also there are vulnerabilities with "unknown" severity, shall it be part of this ADR?

Reply:
Do you have an example of this? Is there a bug tracking this?

Reply:
it's in clair output and also in quay.io 

https://quay.io/repository/martinbasti/single-container-app/manifest/sha256:edbffe44dd94babd6e4d25708cff444e23e7db8eb7c427ef1d0c50ae33d26f68?tab=vulnerabilities&fixable=true

There is a bug tracker, but I cannot get the ID right now

Reply:
Fyi the [STONEINTG-425](https://issues.redhat.com/browse/STONEINTG-425) is now being implemented so the `unknown` severity vulnerabilities will start to show up in this output this week.

Reply:
I added this into these into the ADR.